keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -339,24 +351,301 @@ service CSIProxyService {
+    // The whole disk will be partitioned
+    //
+    // Restrictions:
+    // Disk device needs to follow Win32 format for device names: \\.\PhysicalDriveX
+    // The prefix has to be: \\.\PhysicalDrive and the suffix is an integer in range:
+    // 0 to maximum number drives allowed by Windows OS.
+    string disk_device = 1;
+
+    // Disk partition type
+    enum ParitionType {
+        MBR = 0;
+        GPT = 1;
+    }
+    PartitionType type = 2;
+}
+
+message PartitionDiskResponse {
+    // Volume device resulting from the partition
+    // Volume device will follow Win32 namespaced GUID format for volumes: \\?\Volume\{GUID}
+    // The prefix has to be: \\?\Volume\ and the suffix to be a GUID enclosed with {}
+    string volume_device = 1;
+}
+
+message FormatVolumeRequest {
+    // The Windows volume device to format
+    // Typically Volume Device returned by PartitionDiskResponse
+    //
+    // Restrictions:
+    // Volume device needs to follow Win32 namespaced GUID format for volumes: \\?\Volume\{GUID}
+    // The prefix has to be: \\?\Volume\ and the suffix to be a GUID enclosed with {}
+    string volume_device = 1;
+
+    // FileSystem type
+    enum FileSystemType {
+        NTFS = 0;
+        FAT = 1;
+    }
+    FileSystemType type = 2;
+}
+
+message FormatVolumeResponse {
+    // Intentionally empty.
+}
+
+message MountSMBShareRequest {
+    // A remote SMB share to mount
+    // All unicode characters allowed in SMB server name specifications are
+    // permitted except for restrictions below
+    //
+    // Restrictions:
+    // SMB share specified in the format: \\server-name\sharename, \\server.fqdn\sharename or \\a.b.c.d\sharename
+    // If not an IP address, share name has to be a valid DNS name.
+    // UNC specifications to local paths or prefix: \\?\ is not allowed.
+    // Characters: + [ ] " / : ; | < > , ? * = $ are not allowed.
+    string remote_share = 1;
+
+    // Local path in the host to stage the SMB share.
+    // All special characters allowed by Windows in path names will be allowed
+    // except for restrictions noted below. For details, please check:
+    // https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file
+    //
+    // Restrictions:
+    // Needs to be an absolute path under kubernetes-csi-plugins-path.
+    // Needs to exist already in host
+    // Path needs to be specified with drive letter prefix: "X:\".
+    // UNC paths of the form "\\server\share\path\file" are not allowed.
+    // All directory separators need to be backslash character: "\".
+    // Characters: .. / : | ? * in the path are not allowed.
+    // Maximum path length will be capped to 260 characters (MAX_PATH).
+    string host_path = 2;
+
+    // Mount the share read-only
+    bool readonly = 3;
+
+    // Username credential associated with the share
+    string username = 4;
+
+    // Password credential associated with the share
+    string password = 5;
```

Comment:
are username/password optional? is there any provision for non-password authentication methods in the future?

Reply:
username/password is indeed optional. At the CSI API level, secrets can be arbitrary items in a map. At the "down below" CSI Proxy level, I am trying to keep things as close to the Windows system and powershell interfaces. For SMB, Windows APIs/cmdlets only accept username and password: https://docs.microsoft.com/en-us/powershell/module/smbshare/new-smbmapping?view=win10-ps